<a href="https://colab.research.google.com/github/FareedaAbdelaal/ML-55-9988/blob/main/ML_Task4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, classification_report

In [5]:
from google.colab import files
uploaded = files.upload()

Saving data_banknote_authentication.csv to data_banknote_authentication.csv


In [6]:
df = pd.read_csv("data_banknote_authentication.csv")

df.head()

,Variance_Wavelet,Skewness_Wavelet,Curtosis_Wavelet,Image_Entropy,Class
0,3.62160,8.6661,-2.8073,-0.44699,0
1,4.54590,8.1674,-2.4586,-1.46210,0
2,3.86600,-2.6383,1.9242,0.10645,0
3,3.45660,9.5228,-4.0112,-3.59440,0
4,0.32924,-4.4552,4.5718,-0.98880,0


In [8]:
df.info()
df.describe()
df['Class'].value_counts()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1372 entries, 0 to 1371
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Variance_Wavelet  1372 non-null   float64
 1   Skewness_Wavelet  1372 non-null   float64
 2   Curtosis_Wavelet  1372 non-null   float64
 3   Image_Entropy     1372 non-null   float64
 4   Class             1372 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 53.7 KB


,count
Class,
0,762
1,610


In [11]:
X = df.drop("Class", axis=1)
y = df["Class"]

In [12]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [13]:
dt = DecisionTreeClassifier()

param_grid_dt = {
    'max_depth': [3,5,7,10,None],
    'min_samples_split': [2,5,10],
    'criterion': ['gini','entropy']
}

grid_dt = GridSearchCV(
    dt,
    param_grid_dt,
    cv=5,
    scoring='accuracy'
)

grid_dt.fit(X_train, y_train)

best_dt = grid_dt.best_estimator_

print("Best Decision Tree Parameters:")
print(grid_dt.best_params_)

Best Decision Tree Parameters:
{'criterion': 'gini', 'max_depth': 10, 'min_samples_split': 2}


In [14]:
y_pred_dt = best_dt.predict(X_test)

print("Decision Tree Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_dt))

print("\nDecision Tree Classification Report:")
print(classification_report(y_test, y_pred_dt))

Decision Tree Confusion Matrix:
[[147   1]
 [  5 122]]

Decision Tree Classification Report:
              precision    recall  f1-score   support

           0       0.97      0.99      0.98       148
           1       0.99      0.96      0.98       127

    accuracy                           0.98       275
   macro avg       0.98      0.98      0.98       275
weighted avg       0.98      0.98      0.98       275



In [15]:
rf = RandomForestClassifier()

param_grid_rf = {
    'n_estimators': [50,100,200],
    'max_depth': [None,5,10],
    'min_samples_split': [2,5]
}

grid_rf = GridSearchCV(
    rf,
    param_grid_rf,
    cv=5,
    scoring='accuracy'
)

grid_rf.fit(X_train, y_train)

best_rf = grid_rf.best_estimator_

print("Best Random Forest Parameters:")
print(grid_rf.best_params_)

Best Random Forest Parameters:
{'max_depth': None, 'min_samples_split': 5, 'n_estimators': 100}


In [16]:
y_pred_rf = best_rf.predict(X_test)

print("Random Forest Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf))

print("\nRandom Forest Classification Report:")
print(classification_report(y_test, y_pred_rf))

Random Forest Confusion Matrix:
[[147   1]
 [  2 125]]

Random Forest Classification Report:
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       148
           1       0.99      0.98      0.99       127

    accuracy                           0.99       275
   macro avg       0.99      0.99      0.99       275
weighted avg       0.99      0.99      0.99       275



Model Comparison:

The Random Forest model performd better than the Decision Tree model. Random Forest achieved 99% accuracy compared to 98% for Decision Tree. Because random forest reduced overfitting by combining multiple trees and it improves generalization performance.


